### PR TITLE
Mitigate JsonObject performance regression.

### LINKED ISSUE #107869: Performance regression with JsonObject creation by +70%
### Description

I have a number of .NET applications I maintain, which I've also recently started maintaining some basic continuous benchmarks for. A number of these applications include an endpoints that's used as a health check for deployment/debugging and is implemented something like this:

```csharp
        app.MapGet("/version", static () =>
        {
            return new JsonObject()
            {
                ["applicationVersion"] = GitMetadata.Version,
                ["frameworkDescription"] = RuntimeInformation.FrameworkDescription,
                ["operatingSystem"] = new JsonObject()
                {
                    ["description"] = RuntimeInformation.OSDescription,
                    ["architecture"] = RuntimeInformation.OSArchitecture.ToString(),
                    ["version"] = Environment.OSVersion.VersionString,
                    ["is64Bit"] = Environment.Is64BitOperatingSystem,
                },
                ["process"] = new JsonObject()
                {
                    ["architecture"] = RuntimeInformation.ProcessArchitecture.ToString(),
                    ["is64BitProcess"] = Environment.Is64BitProcess,
                    ["isNativeAoT"] = !RuntimeFeature.IsDynamicCodeSupported,
                    ["isPrivilegedProcess"] = Environment.IsPrivilegedProcess,
                },
                ["dotnetVersions"] = new JsonObject()
                {
                    ["runtime"] = GetVersion<object>(),
                    ["aspNetCore"] = GetVersion<HttpContext>(),
                },
            };

            static string GetVersion<T>()
                => typeof(T).Assembly.GetCustomAttribute<AssemblyInformationalVersionAttribute>()!.InformationalVersion;
        });
```

As it's fairly trivial (e.g. has no external dependencies) I've been including it in the benchmarks as a sort of "canary" as a proxy for the overall performance of ASP.NET Core, JSON, the runtime etc.

I've over the last week updated a number of applications from .NET 8 to .NET 9 RC1 and was surprised to find that I'm consistently seeing a regression from this endpoint in terms of the memory allocated compared to .NET 8.

Below is an example chart from my tracking dashboard. The point at which the memory goes up and the duration goes down is the commit that applied the .NET 9 upgrade:

image

I've pulled this out into its own dedicated benchmark that just creates the `JsonObject`, and I see the same regression in terms of memory usage (and also duration). I've made an assumption here that the regression is in `JsonObject` as the properties being serialized are ones I would have thought wouldn't changed during the lifetime of a process and thus be cached.

Admittedly this isn't a critical path in terms of performance, but as it went backwards against what I would have thought the trend would be for the upgrade, and that it was noticeable, I figured I'd raise an issue so that if it's not an expected trade-off somewhere it can be fixed.

These are the results I get on my laptop for the simplified benchmark:

```

BenchmarkDotNet v0.14.0, Windows 11 (10.0.22621.4169/22H2/2022Update/SunValley2)
12th Gen Intel Core i7-1270P, 1 CPU, 16 logical and 12 physical cores
.NET SDK 9.0.100-rc.1.24452.12
  [Host]     : .NET 8.0.8 (8.0.824.36612), X64 RyuJIT AVX2
  Job-WBWTQA : .NET 8.0.8 (8.0.824.36612), X64 RyuJIT AVX2
  Job-MBMPCJ : .NET 9.0.0 (9.0.24.43107), X64 RyuJIT AVX2


```
| Method           | Runtime  | Mean     | Error    | StdDev   | Median   | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|----------------- |--------- |---------:|---------:|---------:|---------:|------:|--------:|-------:|----------:|------------:|
| CreateJsonObject | .NET 8.0 | 400.3 ns |  8.02 ns | 20.71 ns | 391.1 ns |  1.00 |    0.07 | 0.0067 |   1.24 KB |        1.00 |
| CreateJsonObject | .NET 9.0 | 769.2 ns | 14.13 ns | 19.35 ns | 771.6 ns |  1.93 |    0.11 | 0.0572 |   2.12 KB |        1.70 |

Compared to my higher-level benchmarks, it also appears to be slower than .NET 8 as well - I think it's masked there as there's much more going on and improvements elsewhere outweigh the increase in the runtime to create the object.

### Reproduction Steps

```console
dotnet run -c Release -f net8.0 -- --runtimes net8.0 net9.0
```

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <ServerGarbageCollection>true</ServerGarbageCollection>
    <TargetFrameworks>net9.0;net8.0</TargetFrameworks>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="BenchmarkDotNet" Version="0.14.0" />
  </ItemGroup>
</Project>
```

```csharp
using System.Runtime.CompilerServices;
using System.Runtime.InteropServices;
using System.Text.Json.Nodes;
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;
using BenchmarkDotNet.Diagnosers;

BenchmarkRunner.Run<JsonNodeBenchmarks>(args: args);

[MemoryDiagnoser]
public class JsonNodeBenchmarks
{
    [Benchmark]
    public JsonObject CreateJsonObject()
    {
        return new JsonObject()
        {
            ["applicationVersion"] = "1.2.3",
            ["frameworkDescription"] = RuntimeInformation.FrameworkDescription,
            ["operatingSystem"] = new JsonObject()
            {
                ["description"] = RuntimeInformation.OSDescription,
                ["architecture"] = RuntimeInformation.OSArchitecture.ToString(),
                ["version"] = Environment.OSVersion.VersionString,
                ["is64Bit"] = Environment.Is64BitOperatingSystem,
            },
            ["process"] = new JsonObject()
            {
                ["architecture"] = RuntimeInformation.ProcessArchitecture.ToString(),
                ["is64BitProcess"] = Environment.Is64BitProcess,
                ["isNativeAoT"] = !RuntimeFeature.IsDynamicCodeSupported,
                ["isPrivilegedProcess"] = Environment.IsPrivilegedProcess,
            },
        };
    }
}
```

### Expected behavior

Performance using .NET 9 is on par, or better, with .NET 8.

### Actual behavior

Duration and memory allocations are greater running under .NET 9, by ~+80% for duration and ~+70% for allocations.

### Regression?

Yes.

### Known Workarounds

None.

### Configuration

.NET SDK 9.0.100-rc.1.24452.12.

### Other information

_No response_

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
Makes the following changes:

1. ~Makes the `JsonValueKind` computation for `JsonValue` instances backed by primitives a lazy operation.~
2. Updates the `JsonObject` indexer implementation so that duplicate lookups aren't performed in the common case where a property isn't being overwritten.

Should be backported to .NET 9. Fix #107869.

## Benchmarks

Using https://github.com/dotnet/performance/pull/4463 on arm64.

| Method                  | Branch | Mean     | Error    | StdDev   | Median   | Min      | Max      | Ratio | RatioSD | Gen0   | Allocated | Alloc Ratio |
|------------------------ |------- |---------:|---------:|---------:|---------:|---------:|---------:|------:|--------:|-------:|----------:|------------:|
| Create_JsonBool         | .NET 8 | 266.5 ns |  3.24 ns |  3.03 ns | 266.1 ns | 262.5 ns | 272.6 ns |  1.00 |    0.02 | 0.4779 |    2000 B |        1.00 |
| Create_JsonBool         |     PR | 277.2 ns |  2.87 ns |  2.54 ns | 277.7 ns | 273.5 ns | 282.0 ns |  1.04 |    0.01 | 0.4781 |    2000 B |        1.00 |
|                         |        |          |          |          |          |          |          |       |         |        |           |             |
| Create_JsonNumber       | .NET 8 | 268.0 ns |  3.33 ns |  3.11 ns | 268.9 ns | 260.5 ns | 271.3 ns |  1.00 |    0.02 | 0.4773 |    2000 B |        1.00 |
| Create_JsonNumber       |     PR | 285.9 ns | 15.21 ns | 17.52 ns | 275.6 ns | 269.9 ns | 328.5 ns |  1.07 |    0.06 | 0.5732 |    2400 B |        1.20 |
|                         |        |          |          |          |          |          |          |       |         |        |           |             |
| Create_JsonString       | .NET 8 | 366.3 ns |  3.87 ns |  3.62 ns | 366.9 ns | 361.1 ns | 372.9 ns |  1.00 |    0.01 | 0.5738 |    2400 B |        1.00 |
| Create_JsonString       |     PR | 281.5 ns |  2.38 ns |  1.98 ns | 280.8 ns | 278.4 ns | 286.1 ns |  0.77 |    0.01 | 0.6689 |    2800 B |        1.17 |
|                         |        |          |          |          |          |          |          |       |         |        |           |             |
| Create_JsonArray        | .NET 8 | 486.5 ns |  6.29 ns |  5.89 ns | 486.5 ns | 477.5 ns | 495.4 ns |  1.00 |    0.02 | 0.7263 |    3040 B |        1.00 |
| Create_JsonArray        |     PR | 463.9 ns |  4.96 ns |  4.64 ns | 463.1 ns | 455.2 ns | 471.9 ns |  0.95 |    0.01 | 0.7258 |    3040 B |        1.00 |
|                         |        |          |          |          |          |          |          |       |         |        |           |             |
| Create_JsonObject_Small | .NET 8 | 236.1 ns |  1.31 ns |  1.16 ns | 236.1 ns | 233.7 ns | 237.8 ns |  1.00 |    0.01 | 0.1618 |     680 B |        1.00 |
| Create_JsonObject_Small |     PR | 198.7 ns |  2.13 ns |  1.99 ns | 198.2 ns | 195.9 ns | 202.0 ns |  0.84 |    0.01 | 0.2503 |    1048 B |        1.54 |
|                         |        |          |          |          |          |          |          |       |         |        |           |             |
| Create_JsonObject_Large | .NET 8 | 639.4 ns |  6.81 ns |  6.37 ns | 638.6 ns | 631.1 ns | 652.8 ns |  1.00 |    0.01 | 0.5612 |    2352 B |        1.00 |
| Create_JsonObject_Large |     PR | 393.0 ns |  2.78 ns |  2.60 ns | 393.0 ns | 389.8 ns | 397.8 ns |  0.61 |    0.01 | 0.5102 |    2136 B |        0.91 |

### CODE REVIEW

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs [hunk 1]
```diff
@@ -248,17 +248,21 @@ internal void SetItem(string propertyName, JsonNode? value)
 
             OrderedDictionary<string, JsonNode?> dict = Dictionary;
 
-            if (dict.TryGetValue(propertyName, out JsonNode? replacedValue))
+            if (!dict.TryAdd(propertyName, value))
             {
+                int index = dict.IndexOf(propertyName);
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs:253]: The duplicate lookup for the case of overwrites could be avoided assuming we had a method similar to `CollectionsMarshal.GetValueRefOrAddDefault`. Alternatively, we could consider adding an overload for `TryAdd` that returns the found index as an `out` parameter. Any particular preferences @stephentoub?

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValueOfTPrimitive.cs [hunk 2]
```diff
@@ -13,18 +13,17 @@ namespace System.Text.Json.Nodes
...
         {
             Debug.Assert(TypeIsSupportedPrimitive, $"The type {typeof(TValue)} is not a supported primitive.");
             Debug.Assert(converter is { IsInternalConverter: true, ConverterStrategy: ConverterStrategy.Value });
 
             _converter = converter;
-            _valueKind = DetermineValueKind(value);
         }
 
-        private protected override JsonValueKind GetValueKindCore() => _valueKind;
+        private protected override JsonValueKind GetValueKindCore() => _valueKind ??= DetermineValueKind(Value);
```
Comment (jkotas) [inline src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValueOfTPrimitive.cs:26]: Nullable<T> assignment is not atomic and thread safe. Is it guaranteed that there are no multiple threads running through this in parallel?

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs (hunk 1 above)
Comment (stephentoub) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs:253]: I'd generally prefer a solution that was considered "safe" and that we could put directly onto the type.

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValueOfTPrimitive.cs (hunk 2 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonValueOfTPrimitive.cs:26]: Good catch. The type is mutable and not threadsafe but we should guarantee that concurrent reads are safe.

**Comment by stephentoub** [comment]
> Revert delayed JsonValueKind derivation

How much of the improvement came from this? There are ways to make it safe.

**Comment by eiriktsarpalis** [comment]
> > Revert delayed JsonValueKind derivation
> 
> How much of the improvement came from this? There are ways to make it safe.

Not a huge amount. I figured I would rather take the regression for the sake of simplicity.

File: src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs (hunk 1 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/src/System/Text/Json/Nodes/JsonObject.cs:253]: API proposal: https://github.com/dotnet/runtime/issues/107947

**Review by krwq — APPROVED** [review]

**Pull request merged**